pyproject.toml

```diff
@@ -2,6 +2,134 @@
 requires = ["setuptools", "wheel", "setuptools-rust"]
 build-backend = "setuptools.build_meta"
 
+[project]
+name = "qiskit"
+description = "An open-source SDK for working with quantum computers at the level of extended quantum circuits, operators, and primitives."
+requires-python = ">=3.8"
+license = { file = "LICENSE.txt" }
+authors = [
+    { name = "Qiskit Development Team", email = "hello@qiskit.org" },
+]
+keywords = [
+    "qiskit",
+    "quantum circuit",
+    "quantum computing",
+    "quantum programming language",
+    "quantum",
+    "sdk",
+]
+classifiers = [
+    "Environment :: Console",
+    "Intended Audience :: Developers",
+    "Intended Audience :: Science/Research",
+    "License :: OSI Approved :: Apache Software License",
+    "Operating System :: MacOS",
+    "Operating System :: Microsoft :: Windows",
+    "Operating System :: POSIX :: Linux",
+    "Programming Language :: Python :: 3 :: Only",
+    "Programming Language :: Python :: 3.8",
+    "Programming Language :: Python :: 3.9",
+    "Programming Language :: Python :: 3.10",
+    "Programming Language :: Python :: 3.11",
+    "Topic :: Scientific/Engineering",
+]
+# These are configured in the `tool.setuptools.dynamic` table.
+dynamic = ["version", "readme", "dependencies"]
+
+# If modifying this table, be sure to sync with `requirements-optional.txt`,
+# `qiskit.utils.optionals`, and `qiskit_pkg`'s `pyproject.toml` too.
+[project.optional-dependencies]
+qasm3-import = [
+    "qiskit-qasm3-import >= 0.1.0",
+]
+visualization = [
+    "matplotlib >= 3.3",
+    "ipywidgets >= 7.3.0",
+    "pydot",
+    "Pillow >= 4.2.1",
+    "pylatexenc >= 1.4",
+    "seaborn >= 0.9.0",
+    "pygments >= 2.4",
+]
+crosstalk-pass = [
+    "z3-solver >= 4.7",
+]
+csp-layout-pass = [
+    "python-constraint >= 1.4",
+]
+# This will make the resolution work for installers from PyPI, but `pip install .[all]` will be
+# unreliable because `qiskit-terra` will resolve to the PyPI version, so local changes in the
+# optionals won't be reflected.
+all = ["qiskit[qasm3-import,visualization,crosstalk-pass,csp-layout-pass]"]
```

Comment:
This seems less than ideal, but in practice I guess people aren't using the `all` extra very frequently. I can't think of a better way to do this besides copy and pasting the component lists for `all` too. I'm guessing we can't use `dynamic` for optional dependencies because that's my only other thought.

Reply:
Yeah, it's not super ideal, though I was assuming `.[all]` isn't being used at all.